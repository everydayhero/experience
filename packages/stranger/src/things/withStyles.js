import React, { PropTypes } from 'react'
import stranger from '../'
import stylesToClasses from './stylesToClasses'

const withStyles = (styles = {}) => (Component = 'div') => {
  const WithStylesComponent = (props, { traits }) => {
    const combinedProps = {
      ...Component.defaultProps,
      ...props
    }
    const stylesIsFunction = typeof styles === 'function'
    const stylesObj = stylesIsFunction
      ? styles(combinedProps, traits || stranger.defaultTraits)
      : styles
    return (
      <Component
        {...combinedProps}
        styles={stylesObj}
        classNames={stylesToClasses(stylesObj)}
      />
    )
  }
  WithStylesComponent.withStyles = true
  WithStylesComponent.contextTypes = {
    traits: PropTypes.object
  }
  return WithStylesComponent
}

export default withStyles
