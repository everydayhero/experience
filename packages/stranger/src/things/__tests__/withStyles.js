import React from 'react'
import ReactShallowRenderer from 'react-test-renderer/shallow'
import withStyles from '../withStyles.js'

const shallow = new ReactShallowRenderer()

describe('comp', () => {
  it('should turn a style object into classes and passes down all styles', () => {
    const styles = ({ props: { makeit = 'red' } }) => ({
      root: {
        color: makeit
      },
      child: {
        backgroundColor: makeit
      }
    })
    const MyComponent = ({
      children,
      classNames
    }) => (
      <div className={classNames.root}>
        <span className={classNames.child}>{children}</span>
      </div>
    )
    const ComponentWithStyles = withStyles(styles)(MyComponent)
    const actual = shallow.render(
      <ComponentWithStyles makeit='green' />
    )
    expect(actual.props.classNames).toMatchSnapshot()
    expect(actual.props.styles).toMatchSnapshot()
  })

  it('should make traits available to style objects', () => {
    const styles = ({ traits: { colors } }) => ({
      root: {
        color: colors.green
      },
      child: {
        backgroundColor: colors.theme.soft
      }
    })
    const MyComponent = ({
      children,
      classNames
    }) => (
      <div className={classNames.root}>
        <span className={classNames.child}>{children}</span>
      </div>
    )
    const ComponentWithStyles = withStyles(styles)(MyComponent)
    const actual = shallow.render(
      <ComponentWithStyles />
    )
    expect(actual.props.classNames).toMatchSnapshot()
    expect(actual.props.styles).toMatchSnapshot()
  })
})
