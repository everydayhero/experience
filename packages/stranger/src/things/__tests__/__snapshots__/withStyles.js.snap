exports[`comp should make traits available to style objects 1`] = `
Object {
  "child": "bc-rgb255_255_255",
  "root": "c-rgb27_171_107",
}
`;

exports[`comp should make traits available to style objects 2`] = `
Object {
  "child": Object {
    "backgroundColor": "rgb(255,255,255)",
  },
  "root": Object {
    "color": "rgb(27,171,107)",
  },
}
`;

exports[`comp should turn a style object into classes and passes down all styles 1`] = `
Object {
  "child": "bc-green",
  "root": "c-green",
}
`;

exports[`comp should turn a style object into classes and passes down all styles 2`] = `
Object {
  "child": Object {
    "backgroundColor": "green",
  },
  "root": Object {
    "color": "green",
  },
}
`;
